### PR TITLE
fix(cli): jwt errors

### PR DESCRIPTION
**Pull request by 7ttp**
## TL;DR

standardizes JWT admission failures across legacy and next local runtimes with 
`{ code, message }` and `sb-error-code` while preserving legacy `msg` for BC

## Ref

- Paired with https://github.com/supabase/supabase/pull/47920
 & https://github.com/supabase/supabase/pull/48012

- Related to https://github.com/supabase/cli/issues/5861 & https://github.com/supabase/supabase/issues/47835

### CODE REVIEW

**Comment by 7ttp** [comment]
@codex review

**Comment by 7ttp** [comment]
> 👋 Thanks! The linked issue hasn't been marked `open-for-contribution` yet, so this pull request is being closed automatically.
> 
> A maintainer adds that label once an issue is triaged and ready to be worked on. Please wait for the label before opening a pull request. See [CONTRIBUTING.md](../blob/develop/CONTRIBUTING.md) for the full workflow. Once a maintainer adds the `open-for-contribution` label to a linked open issue, reopen or open a new pull request and it will be accepted.

imo  this shouldn't really happen for the members ? 😅

thoughts @avallete @jgoux ?

**Review by kallebysantos — APPROVED** [review]

**Comment by avallete** [comment]
> > 👋 Thanks! The linked issue hasn't been marked `open-for-contribution` yet, so this pull request is being closed automatically.
> > A maintainer adds that label once an issue is triaged and ready to be worked on. Please wait for the label before opening a pull request. See [CONTRIBUTING.md](../blob/develop/CONTRIBUTING.md) for the full workflow. Once a maintainer adds the `open-for-contribution` label to a linked open issue, reopen or open a new pull request and it will be accepted.
> 
> imo this shouldn't really happen for the members ? 😅
> 
> thoughts @avallete @jgoux ?

It shouldn't, sadly if your affiliation is "private" on https://github.com/orgs/supabase/people
Then, since the GH token for the check is a public one, that won't be detected by the action. Then, the fallback is to look for active repo permissions (some admin are private members) but that doesn't apply in your case either 😅 

If that's not a bother, could you try to your supabase membership to "public" ? This should allows your PR's to pass the gate without issues.

**Review by avallete — APPROVED** [review]

**Review by avallete — DISMISSED** [review]
I think this might actually be a BC need to double check.

**Comment by avallete** [comment]
Alright, tell me if I'm mistaken @kallebysantos @7ttp  but the Hosted platform does not uses `{ error }` at all does they ?

Looking at the code, I think this is the wrong fix, and we actually need a contract between edge-functions/stdio/cli for error codes, they seems to solely rely on `{ code, message }` shape. This seems to introduce one more schema rather than align with platform:

```
   Runtime                     JWT error body
  ━━━━━━━━━━━━━━━━━━━━━━━━━━  ━━━━━━━━━━━━━━━━━━━
   Hosted platform             { code, message }
  ──────────────────────────  ───────────────────
   Stable local CLI today      { msg }
  ──────────────────────────  ───────────────────
   PR #5862                    { error }
  ──────────────────────────  ───────────────────
   Next/alpha local stack      { msg }
  ──────────────────────────  ───────────────────
   Self-hosted Docker today    { msg }
```

The official docs shows "error, message" contract: https://github.com/supabase/supabase/blob/master/apps/docs/content/troubleshooting/edge-function-401-error-response.mdx
   
And it look like the only place where we rely on `error` in studio is there: https://github.com/supabase/supabase/blob/master/apps/studio/pages/api/edge-functions/test.ts#L95-L100

I think we should standardize toward `{ code, message }` instead of adding another contract ?
I think the root cause of the issue is in studio:


```mermaid
flowchart LR
    UI["Studio Test UI"]
    SP["Studio test proxy<br/>currently assumes body.error"]
    GW["Gateway<br/>Kong / Envoy / hosted router"]
    JW["JWT admission layer<br/>hosted ingress / local main worker"]
    UF["User function"]

    UI --> SP --> GW --> JW --> UF

    GW -. "503: {message} or text" .-> SP
    JW -. "401: hosted {code,message}<br/>local legacy {msg}" .-> SP
    UF -. "arbitrary status/body/headers" .-> SP
```

The bug I think is https://github.com/supabase/supabase/blob/master/apps/studio/pages/api/edge-functions/test.ts#L92-L107 :

1. It receives a legitimate non-2xx response.
2. It discards the body and headers.
3. It tries only errorBody.error.
4. Everything else becomes “Edge function returned an error.”

That assumption is impossible to satisfy universally:

- Kong can generate {"message":"name resolution failed"} before the runtime. The exact issue reproduction proves this.
- Envoy can generate plain text.
- Hosted ingress emits {code,message} plus sb-error-code (https://github.com/supabase/supabase/blob/master/apps/docs/content/guides/functions/error-codes.mdx).
- Self-hosted JWT handling currently emits {msg} in https://github.com/supabase/supabase/blob/master/docker/volumes/functions/main/index.ts#L111 .
- Edge Runtime itself independently emits {msg} for some failures in https://github.com/supabase/edge-runtime/blob/main/crates/base/src/server.rs#L263-L280
- User functions are allowed to return any HTTP body.

**I think a minimal cleaner fix**:

I would make the Studio test endpoint a transparent response adapter.

A completed upstream request even if it returns 401, 404, or 503 is a successful execution of the test operation. It should return:

```
{
  status: upstream.status,
  headers: upstreamHeaders,
  body: upstreamBody
}
```

The Studio proxy itself should return HTTP 200 for that envelope. Reserve proxy-level 4xx/5xx for cases where Studio could not execute the test at all: invalid URL, network failure, malformed request, etc.

Concretely, in `supabase/apps/studio/pages/api/edge-functions/test.ts`:

- always use await `response.text()` do not parse and reserialize based on `Content-Type`.
- always collect upstream headers.
- remove the special `!response.ok` schema extraction.
- return the same `{status, headers, body}` envelope for every upstream status.

This fits the existing architecture unusually well:

- The mutation already treats a 2xx proxy response as ResponseData in https://github.com/supabase/supabase/blob/master/apps/studio/data/edge-functions/edge-function-test-mutation.ts#L21-L43.
- The UI already knows how to display a destructive status badge, Body tab, and Headers tab for status >= 400 in https://github.com/supabase/supabase/blob/master/apps/studio/components/interfaces/Functions/EdgeFunctionDetails/EdgeFunctionTesterSheet.tsx#L406.
- sb-error-code would become visible instead of being discarded.
- It handles future and arbitrary formats without teaching Studio another list of guessed keys.

#47855 (https://github.com/supabase/supabase/pull/47855)’s tolerant extractor is an acceptable hotfix, but it remains schema guessing and still loses response headers and structured bodies. Transparent relay is cleaner.

Tests should cover hosted {code,message}, Kong {message}, legacy {msg}, nested JSON, arbitrary JSON, plain text, invalid JSON, headers, and a genuine fetch exception.


## Where JWT mirror code should live

JWT errors belong at the layer that performs JWT admission not in Studio and not in the generic Edge Runtime engine.

The canonical contract should be the hosted ingress contract:

```
HTTP/1.1 401 Unauthorized
Content-Type: application/json
sb-error-code: UNAUTHORIZED_NO_AUTH_HEADER

{
  "code": "UNAUTHORIZED_NO_AUTH_HEADER",
  "message": "Missing authorization header"
}
```

Canonical codes:

- UNAUTHORIZED_NO_AUTH_HEADER
- UNAUTHORIZED_INVALID_JWT_FORMAT
- UNAUTHORIZED_LEGACY_JWT
- UNAUTHORIZED_ASYMMETRIC_JWT
- UNAUTHORIZED_UNSUPPORTED_TOKEN_ALGORITHM

The mirrors should live beside JWT verification in each independently shipped bootstrap:

- Hosted authority: edge-functions-ingress/src/main/errors.ts, jwt.ts, and utils.ts.
- Self-hosted Docker: supabase/docker/volumes/functions/main/index.ts.
- Stable CLI runtime: cli/apps/cli/src/shared/functions/serve.main.ts.
- Next/alpha CLI runtime: cli/packages/stack/src/services/edge-runtime-main.ts.

Each bootstrap should have a small local JwtFailure mapping and one unauthorized(failure) response helper.


## Backward-compatible migration

Local {msg} has shipped since 2023, so replace it additively:
```
{
  "code": "UNAUTHORIZED_INVALID_JWT_FORMAT",
  "message": "Invalid JWT format",
  "msg": "Invalid JWT format"
}
```
Also emit `sb-error-code`. Treat msg as deprecated and remove it only in an explicit major-version change or keep the tiny alias indefinitely.

Recommended sequence:

1. Fix Studio to preserve arbitrary upstream responses.
2. Add canonical `{code,message}` and `sb-error-code` to all JWT admission bootstraps.
3. Keep legacy `msg` during migration.
4. Add identical response-contract tests in self-hosted, stable CLI, and next CLI.
5. Re-scope #5862 away from msg → error, to `{code,message}` it currently also misses the next CLI path.

The important boundary is to standardize all Supabase-owned JWT errors, preserve all other HTTP responses transparently.

Trying to standardize gateway failures and user-function responses would require rewriting responses that Supabase does not own.

cc @SaxonF @joshenlim since it seems you authored: https://github.com/supabase/supabase/blame/master/apps/studio/pages/api/edge-functions/test.ts#L92-L107 would you give it a look, tell me if you think I'm off ?

**Comment by kallebysantos** [comment]
You're right... I missed it, and I agree that we should use a single pattern across all environments.
As addition to @avallete comment: https://supabase.com/docs/guides/functions/error-codes

We've been refactoring Platform error codes, and yes! we're now using `{code, message}` object, example:
```ts
export enum AuthErrors {
  UnsupportedTokenAlgorithm = 'UNAUTHORIZED_UNSUPPORTED_TOKEN_ALGORITHM',
  MissingAuthHeader = 'UNAUTHORIZED_NO_AUTH_HEADER',
  // ....
}
```
```ts
// in case JWT validation fails:
const error = {
    code: AuthErrors.UnsupportedTokenAlgorithm,
    message: `Unsupported JWT algorithm ${jwt.alg}`
}

return Response.json(error, { 
  status:  Status.Unauthorized,
  headers: {
    'sb-error-code': AuthErrors.UnsupportedTokenAlgorithm
  }
})
```

File: apps/cli/src/shared/functions/serve.main.ts [hunk 1]
```diff
@@ -245,11 +255,16 @@ Deno.serve({
         const isValidJWT = await verifyHybridJWT(JWT_SECRET, JWKS_ENDPOINT, token);
 
         if (!isValidJWT) {
-          return getResponse({ msg: "Invalid JWT" }, STATUS_CODE.Unauthorized);
+          return getAuthErrorResponse(AUTH_ERROR_CODE.InvalidJwtFormat, "Invalid JWT");
```
[the remark this answers is not in the document]
Comment (7ttp) [reply apps/cli/src/shared/functions/serve.main.ts]: updated this so malformed tokens, 
HS256 failures, asymmetric failures, and unsupported algorithms each return the appropriate error code...

File: apps/cli/src/shared/functions/serve.main.ts [hunk 2]
```diff
@@ -74,6 +109,12 @@ function getResponse(payload: any, status: number, customHeaders = {}) {
   return new Response(body, { status, headers });
 }
 
+function getAuthErrorResponse({ code, message }: AuthFailure) {
+  return getResponse({ code, message, msg: message }, STATUS_CODE.Unauthorized, {
```
[the remark this answers is not in the document]
Comment (7ttp) [reply apps/cli/src/shared/functions/serve.main.ts]: Keeping `{ code, message, msg }` to match the pattern and preserve compatibility 
the `.error` dependency is specific to Studio’s test proxy and should be fixed there

**Comment by 7ttp** [comment]
@codex review

File: apps/cli/src/shared/functions/serve.main.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (7ttp) [reply apps/cli/src/shared/functions/serve.main.ts]: intentional: `{ code, message }` is the the platform way 
`msg` preserves BC , and Studio’s `errorBody.error` assumption should be fixed in its response proxy..

**Comment by 7ttp** [comment]
mb! @avallete @kallebysantos

I’ve rescoped this PR to the `{ code, message }` pattern   (kept msg for now for BC)

 but b4 this will patch the studio test proxy so it preserves the upstream status, headers, and body instead of keying only off `errorBody.error` (waiting for the team's take on it)

**Comment by avallete** [comment]
- PR does not yet resolve/close #5861. The response intentionally uses { code, message, msg }, not error (source (https://github.com/supabase/cli/blob/7cfdaeed5a2d573730e28a03148525ce872ab2f8/apps/cli/src/shared/functions/serve.main.ts#L112-L116)). That aligns with the platform contract, but Studio still reads error, both Studio fixes remain open. Merge one first, or change Closes #5861 and update the stale PR description, which still claims this adds error.
- P2 - next/alpha local runtime remains inconsistent. packages/stack/src/services/edge-runtime-main.ts:63 still returns only { msg }, supabase functions dev therefore lacks { code, message } and sb-error-code.
- P2 - no behavioral coverage. Five externally visible JWT classifications were added without tests asserting status, body, or sb-error-code. Existing tests only cover bundling, artifact drift, and offline startup. Add focused contract cases for missing/malformed auth, invalid HS256/asymmetric JWTs, and unsupported algorithms.

File: packages/stack/src/services/edge-runtime-main.ts [hunk 3]
```diff
@@ -36,15 +72,64 @@ function bytesEqual(left: Uint8Array, right: Uint8Array) {
...
+      status: 401,
+      headers: { "sb-error-code": code },
+    },
+  );
+}
+
+function extractBearerToken(rawToken: string) {
+  const tokenParts = rawToken.split(" ");
+  const [bearer, token] = tokenParts;
+  if (bearer !== "Bearer" || tokenParts.length !== 2) {
```
[the remark this answers is not in the document]
Comment (7ttp) [reply packages/stack/src/services/edge-runtime-main.ts]: https://github.com/supabase/cli/pull/5862/commits/d6b122971aa26847a96a79e92c0817309f003d4d

File: packages/stack/src/services/edge-runtime-main.ts [hunk 4]
```diff
@@ -36,6 +72,25 @@ function bytesEqual(left: Uint8Array, right: Uint8Array) {
   return result === 0;
 }
 
+function getAuthErrorResponse({ code, message }: AuthFailure) {
+  return Response.json(
+    { code, message, msg: message },
+    {
+      status: 401,
+      headers: { "sb-error-code": code },
```
[the remark this answers is not in the document]
Comment (7ttp) [reply packages/stack/src/services/edge-runtime-main.ts]: fixed in https://github.com/supabase/cli/pull/5862/commits/ea3a67c1d41938f81ccf8807830bba865fa591cb

**Comment by 7ttp** [comment]
> * PR does not yet resolve/close [Edge function JWT verification errors return { msg } instead of { error } in serve.main.ts #5861](https://github.com/supabase/cli/issues/5861). The response intentionally uses { code, message, msg }, not error (source (https://github.com/supabase/cli/blob/7cfdaeed5a2d573730e28a03148525ce872ab2f8/apps/cli/src/shared/functions/serve.main.ts#L112-L116
>   )). That aligns with the platform contract, but Studio still reads error, both Studio fixes remain open. Merge one first, or change Closes [Edge function JWT verification errors return { msg } instead of { error } in serve.main.ts #5861](https://github.com/supabase/cli/issues/5861) and update the stale PR description, which still claims this adds error.
> * P2 - next/alpha local runtime remains inconsistent. packages/stack/src/services/edge-runtime-main.ts:63 still returns only { msg }, supabase functions dev therefore lacks { code, message } and sb-error-code.
> * P2 - no behavioral coverage. Five externally visible JWT classifications were added without tests asserting status, body, or sb-error-code. Existing tests only cover bundling, artifact drift, and offline startup. Add focused contract cases for missing/malformed auth, invalid HS256/asymmetric JWTs, and unsupported algorithms.

Thanks! 💚 Addressed all three: updated the desc  and left https://github.com/supabase/cli/issues/5861 related/pending 
- supabase/supabase#47920

& also aligned the next runtime, added behavioral coverage around it...

**Comment by avallete** [comment]
Thank's for the update ! A one more finding that I think need to be addressed before merging:

[Legacy Kong CORS config](https://github.com/supabase/cli/blob/c01411fc3088f59370751b535a82a4605df3d20f/apps/cli-go/internal/start/templates/kong.yml#L168-L181) (https://github.com/supabase/cli/blob/c01411fc3088f59370751b535a82a4605df3d20f/apps/cli-go/internal/start/templates/kong.yml#L168-L181) does not expose sb-error-code. Browser clients cannot read
     the new header. The E2E tests bypasses Kong, so it misses this.

**Comment by 7ttp** [comment]
> Thank's for the update ! A one more finding that I think need to be addressed before merging:
> 
> [Legacy Kong CORS config](https://github.com/supabase/cli/blob/c01411fc3088f59370751b535a82a4605df3d20f/apps/cli-go/internal/start/templates/kong.yml#L168-L181) (
> 
> https://github.com/supabase/cli/blob/c01411fc3088f59370751b535a82a4605df3d20f/apps/cli-go/internal/start/templates/kong.yml#L168-L181
> 
> ) does not expose sb-error-code. Browser clients cannot read
> the new header. The E2E tests bypasses Kong, so it misses this.

addressed in the latest commit..

File: apps/cli-go/internal/start/templates/kong.yml [hunk 5]
```diff
@@ -178,6 +178,9 @@ services:
           - /functions/v1/
     plugins:
       - name: cors
+        config:
+          exposed_headers:
+            - sb-error-code
```
Comment (avallete) [inline apps/cli-go/internal/start/templates/kong.yml]: **issue**

Thanks for exposing sb-error-code, but configuring it globally in Kong introduces a CORS regression (found out while dogfooding locally).

  `Access-Control-Expose-Headers` is the browser’s permission list for response headers. A user function may return:
```
  X-Custom-Id: abc123
  Access-Control-Expose-Headers: X-Custom-Id
```
  This allows browser code to read:
```
  response.headers.get("X-Custom-Id")
```
  With:
```
  config:
    exposed_headers:
      - sb-error-code
```
  Kong uses set_header, not append/merge. It replaces the upstream value, producing:
```
  X-Custom-Id: abc123
  Access-Control-Expose-Headers: sb-error-code
```
  `X-Custom-Id` still travels over the network, but browser JavaScript can no longer read it. Because this CORS plugin applies to the entire Functions service, it affects successful and user-defined responses not only Supabase JWT errors.

  A safer fix is to remove the global Kong exposed_headers configuration and add: `Access-Control-Expose-Headers: sb-error-code` directly to the runtime-owned JWT error response. That exposes sb-error-code only where needed while preserving user functions’ CORS headers.

src/shared/functions/serve.main.ts, inside getAuthErrorResponse:
```
  function getAuthErrorResponse({ code, message }: AuthFailure) {
    return getResponse({ code, message, msg: message }, STATUS_CODE.Unauthorized, {
      "sb-error-code": code,
      "Access-Control-Expose-Headers": "sb-error-code",
    });
  }
```
This confines the exposure permission to Supabase-generated JWT errors.

  Could we also add a behavioral regression test through Kong? The current auth E2E calls edge-runtime directly, so it cannot detect this. The test should verify:

  1. A user-function response with Access-Control-Expose-Headers: X-Custom still exposes X-Custom after passing through Kong.
  2. A JWT rejection passing through Kong exposes sb-error-code and allows browser clients to read it.

  That would protect both the new error header and existing user-defined headers from future regressions, since it's a tricky one to spot.

Comment (7ttp) [reply apps/cli-go/internal/start/templates/kong.yml]: sorted in: https://github.com/supabase/cli/pull/5862/commits/2c5740e38586f0cf05972f155fa8d39965ab4ac2

**Review by Coly010 — APPROVED** [review]
